File: backend/api-server/src/processors/DocumentProcessor.js
import fs from 'fs-extra';
import path from 'path';
import mammoth from 'mammoth';
import XLSX from 'xlsx';
import csv from 'csv-parser';
import { Readable } from 'stream';

class DocumentProcessor {
  constructor() {    this.supportedTypes = {
      'text/plain': this.processTextFile.bind(this),
      'text/markdown': this.processTextFile.bind(this),
      'application/msword': this.processDocFile.bind(this),
      'application/vnd.openxmlformats-officedocument.wordprocessingml.document': this.processDocxFile.bind(this),
      'application/vnd.ms-excel': this.processExcelFile.bind(this),
      'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet': this.processExcelFile.bind(this),
      'text/csv': this.processCSVFile.bind(this),
      'application/json': this.processJSONFile.bind(this)
    };
  }

  async processFile(filePath, mimeType, originalName) {
    try {
      // Check by MIME type first
      if (this.supportedTypes[mimeType]) {
        return await this.supportedTypes[mimeType](filePath, originalName);
      }      // Fallback to file extension
      const ext = path.extname(originalName).toLowerCase();
      switch (ext) {
        case '.txt':
        case '.md':
          return await this.processTextFile(filePath, originalName);
        case '.doc':
          return await this.processDocFile(filePath, originalName);
        case '.docx':
          return await this.processDocxFile(filePath, originalName);
        case '.xls':
        case '.xlsx':
          return await this.processExcelFile(filePath, originalName);
        case '.csv':
          return await this.processCSVFile(filePath, originalName);
        case '.json':
          return await this.processJSONFile(filePath, originalName);
        default:
          throw new Error(`Unsupported file type: ${ext}. PDF support coming in Phase 2!`);
      }
    } catch (error) {
      console.error(`Error processing file ${originalName}:`, error);
      throw new Error(`Failed to process ${originalName}: ${error.message}`);
    }
  }

  async processTextFile(filePath, originalName) {
    try {
      const content = await fs.readFile(filePath, 'utf-8');
      return this.cleanText(content);
    } catch (error) {
      throw new Error(`Failed to read text file: ${error.message}`);
    }
  }
  async processPDF(filePath, originalName) {
    throw new Error('PDF support is coming in Phase 2! Please use TXT, DOCX, CSV, XLS, or JSON files for now.');
  }

  async processDocFile(filePath, originalName) {
    throw new Error('Legacy .doc files are not supported in this version. Please convert to .docx format.');
  }

  async processDocxFile(filePath, originalName) {
    try {
      const result = await mammoth.extractRawText({ path: filePath });
      if (result.messages.length > 0) {
        console.warn('Mammoth warnings for', originalName, ':', result.messages);
      }
      return this.cleanText(result.value);
    } catch (error) {
      throw new Error(`Failed to parse DOCX: ${error.message}`);
    }
  }

  async processExcelFile(filePath, originalName) {
    try {
      const workbook = XLSX.readFile(filePath);
      let allText = '';
      
      // Process all sheets
      workbook.SheetNames.forEach(sheetName => {
        const worksheet = workbook.Sheets[sheetName];
        const jsonData = XLSX.utils.sheet_to_json(worksheet, { header: 1 });
        
        // Add sheet name as context
        allText += `\\n\\n=== Sheet: ${sheetName} ===\\n`;
        
        // Convert rows to text
        jsonData.forEach((row, rowIndex) => {
          if (row.length > 0) {
            const rowText = row.map(cell => 
              cell !== null && cell !== undefined ? String(cell).trim() : ''
            ).filter(cell => cell.length > 0).join(' | ');
            
            if (rowText.length > 0) {
              allText += `Row ${rowIndex + 1}: ${rowText}\\n`;
            }
          }
        });
      });

      return this.cleanText(allText);
    } catch (error) {
      throw new Error(`Failed to parse Excel file: ${error.message}`);
    }
  }

  async processCSVFile(filePath, originalName) {
    try {
      const csvData = [];
      const fileContent = await fs.readFile(filePath, 'utf-8');
      
      return new Promise((resolve, reject) => {
        const stream = Readable.from([fileContent]);
        
        stream
          .pipe(csv())
          .on('data', (row) => {
            csvData.push(row);
          })
          .on('end', () => {
            try {
              let text = '';
              
              // Add headers if available
              if (csvData.length > 0) {
                const headers = Object.keys(csvData[0]);
                text += `Headers: ${headers.join(' | ')}\\n\\n`;
                
                // Add each row
                csvData.forEach((row, index) => {
                  const rowText = headers.map(header => 
                    row[header] !== null && row[header] !== undefined ? String(row[header]).trim() : ''
                  ).filter(cell => cell.length > 0).join(' | ');
                  
                  if (rowText.length > 0) {
                    text += `Row ${index + 1}: ${rowText}\\n`;
                  }
                });
              }
              
              resolve(this.cleanText(text));
            } catch (error) {
              reject(new Error(`Failed to process CSV data: ${error.message}`));
            }
          })
          .on('error', (error) => {
            reject(new Error(`Failed to parse CSV: ${error.message}`));
          });
      });
    } catch (error) {
      throw new Error(`Failed to read CSV file: ${error.message}`);
    }
  }

  async processJSONFile(filePath, originalName) {
    try {
      const content = await fs.readFile(filePath, 'utf-8');
      const jsonData = JSON.parse(content);
      
      // Convert JSON to searchable text
      const text = this.jsonToText(jsonData);
      return this.cleanText(text);
    } catch (error) {
      throw new Error(`Failed to parse JSON: ${error.message}`);
    }
  }

  jsonToText(obj, prefix = '') {
    let text = '';
    
    if (Array.isArray(obj)) {
      obj.forEach((item, index) => {
        text += `${prefix}[${index}]: ${this.jsonToText(item, prefix + '  ')}\\n`;
      });
    } else if (typeof obj === 'object' && obj !== null) {
      Object.entries(obj).forEach(([key, value]) => {
        if (typeof value === 'object') {
          text += `${prefix}${key}:\\n${this.jsonToText(value, prefix + '  ')}`;
        } else {
          text += `${prefix}${key}: ${String(value)}\\n`;
        }
      });
    } else {
      text += String(obj);
    }
    
    return text;
  }

  cleanText(text) {
    if (!text) return '';
    
    return text
      // Normalize whitespace
      .replace(/\\s+/g, ' ')
      // Remove excessive newlines
      .replace(/\\n\\s*\\n\\s*\\n+/g, '\\n\\n')
      // Trim
      .trim();
  }
  getSupportedExtensions() {
    return ['.txt', '.md', '.docx', '.xls', '.xlsx', '.csv', '.json'];
  }

  getSupportedMimeTypes() {
    return Object.keys(this.supportedTypes);
  }
}

export default DocumentProcessor;

File: backend/ai-service/app/document_processor.py
import asyncio
from typing import BinaryIO, Dict, Any
import mimetypes
import pandas as pd
from docx import Document as DocxDocument
import json
import logging

logger = logging.getLogger(__name__)

class DocumentProcessor:
    """Process various document types for AI embedding"""
    
    def __init__(self):
        self.supported_types = {
            'text/plain': self._process_text,
            'text/markdown': self._process_text,
            'text/csv': self._process_csv,
            'application/json': self._process_json,
            'application/vnd.openxmlformats-officedocument.wordprocessingml.document': self._process_docx,
            'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet': self._process_excel,
        }
    
    async def process_file(self, content: bytes, filename: str) -> str:
        """Process file content and return clean text for embedding"""
        # Detect MIME type
        mime_type, _ = mimetypes.guess_type(filename)
        
        if not mime_type or mime_type not in self.supported_types:
            # Fallback to extension-based detection
            ext = filename.lower().split('.')[-1]
            mime_type = self._get_mime_from_extension(ext)
        
        if mime_type not in self.supported_types:
            raise ValueError(f"Unsupported file type: {mime_type}")
        
        try:
            # Process the content
            text = await self.supported_types[mime_type](content, filename)
            
            # Clean and normalize the text
            cleaned_text = self._clean_text(text)
            
            logger.info(f"Processed {filename}: {len(cleaned_text)} characters")
            return cleaned_text
            
        except Exception as e:
            logger.error(f"Error processing {filename}: {str(e)}")
            raise
    
    def _get_mime_from_extension(self, ext: str) -> str:
        """Map file extensions to MIME types"""
        mapping = {
            'txt': 'text/plain',
            'md': 'text/markdown',
            'csv': 'text/csv',
            'json': 'application/json',
            'docx': 'application/vnd.openxmlformats-officedocument.wordprocessingml.document',
            'xlsx': 'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet',
        }
        return mapping.get(ext, 'text/plain')
    
    async def _process_text(self, content: bytes, filename: str) -> str:
        """Process plain text files"""
        try:
            # Try UTF-8 first, fallback to latin-1
            try:
                return content.decode('utf-8')
            except UnicodeDecodeError:
                return content.decode('latin-1')
        except Exception as e:
            raise ValueError(f"Failed to decode text file: {str(e)}")
    
    async def _process_csv(self, content: bytes, filename: str) -> str:
        """Process CSV files into structured text"""
        try:
            # Decode content
            text_content = content.decode('utf-8')
            
            # Parse CSV
            from io import StringIO
            df = pd.read_csv(StringIO(text_content))
            
            # Convert to structured text
            result = f"CSV File: {filename}\\n\\n"
            result += f"Columns: {', '.join(df.columns)}\\n\\n"
            
            # Add each row as structured text
            for idx, row in df.iterrows():
                row_text = " | ".join([f"{col}: {val}" for col, val in row.items() if pd.notna(val)])
                result += f"Row {idx + 1}: {row_text}\\n"
            
            return result
            
        except Exception as e:
            raise ValueError(f"Failed to process CSV file: {str(e)}")
    
    async def _process_json(self, content: bytes, filename: str) -> str:
        """Process JSON files into readable text"""
        try:
            text_content = content.decode('utf-8')
            data = json.loads(text_content)
            
            # Convert JSON to structured text
            result = f"JSON File: {filename}\\n\\n"
            result += self._json_to_text(data)
            
            return result
            
        except Exception as e:
            raise ValueError(f"Failed to process JSON file: {str(e)}")
    
    async def _process_docx(self, content: bytes, filename: str) -> str:
        """Process DOCX files"""
        try:
            from io import BytesIO
            doc = DocxDocument(BytesIO(content))
            
            text_parts = []
            for paragraph in doc.paragraphs:
                if paragraph.text.strip():
                    text_parts.append(paragraph.text.strip())
            
            return "\\n".join(text_parts)
            
        except Exception as e:
            raise ValueError(f"Failed to process DOCX file: {str(e)}")
    
    async def _process_excel(self, content: bytes, filename: str) -> str:
        """Process Excel files"""
        try:
            from io import BytesIO
            
            # Read all sheets
            excel_data = pd.read_excel(BytesIO(content), sheet_name=None)
            
            result = f"Excel File: {filename}\\n\\n"
            
            for sheet_name, df in excel_data.items():
                result += f"Sheet: {sheet_name}\\n"
                result += f"Columns: {', '.join(df.columns)}\\n\\n"
                
                # Add each row
                for idx, row in df.iterrows():
                    row_text = " | ".join([f"{col}: {val}" for col, val in row.items() if pd.notna(val)])
                    result += f"Row {idx + 1}: {row_text}\\n"
                
                result += "\\n"
            
            return result
            
        except Exception as e:
            raise ValueError(f"Failed to process Excel file: {str(e)}")
    
    def _json_to_text(self, obj, indent=0) -> str:
        """Convert JSON object to readable text"""
        result = ""
        prefix = "  " * indent
        
        if isinstance(obj, dict):
            for key, value in obj.items():
                if isinstance(value, (dict, list)):
                    result += f"{prefix}{key}:\\n"
                    result += self._json_to_text(value, indent + 1)
                else:
                    result += f"{prefix}{key}: {value}\\n"
        elif isinstance(obj, list):
            for i, item in enumerate(obj):
                result += f"{prefix}Item {i + 1}:\\n"
                result += self._json_to_text(item, indent + 1)
        else:
            result += f"{prefix}{obj}\\n"
        
        return result
    
    def _clean_text(self, text: str) -> str:
        """Clean and normalize text for embedding"""
        # Remove excessive whitespace
        text = " ".join(text.split())
        
        # Remove very short lines that might be noise
        lines = text.split('\\n')
        cleaned_lines = [line.strip() for line in lines if len(line.strip()) > 2]
        
        return "\\n".join(cleaned_lines)
